docs/eBpfExtensions.md

```diff
@@ -1,42 +1,207 @@
+
+#### `ebpf_argument_type_t` Enum
+This enum describes the various argument types that can be passed to an eBPF helper function. This is defined in [Prevail Verifier](https://github.com/vbpf/ebpf-verifier) project.
+```
+typedef enum _ebpf_argument_type {
+    EBPF_ARGUMENT_TYPE_DONTCARE = 0,
+    EBPF_ARGUMENT_TYPE_ANYTHING,
+    EBPF_ARGUMENT_TYPE_CONST_SIZE,
+    EBPF_ARGUMENT_TYPE_CONST_SIZE_OR_ZERO,
+    EBPF_ARGUMENT_TYPE_PTR_TO_CTX,
+    EBPF_ARGUMENT_TYPE_PTR_TO_MAP,
+    EBPF_ARGUMENT_TYPE_PTR_TO_MAP_OF_PROGRAMS,
+    EBPF_ARGUMENT_TYPE_PTR_TO_MAP_KEY,
+    EBPF_ARGUMENT_TYPE_PTR_TO_MAP_VALUE,
+    EBPF_ARGUMENT_TYPE_PTR_TO_MEM,
+    EBPF_ARGUMENT_TYPE_PTR_TO_MEM_OR_NULL,
+    EBPF_ARGUMENT_TYPE_PTR_TO_UNINIT_MEM,
+} ebpf_argument_type_t;
+```
+
+#### `ebpf_return_type` Enum
+This enum describes the various return types from an eBPF helper function. This is defined in [Prevail Verifier](https://github.com/vbpf/ebpf-verifier) project.
+```
+typedef enum _ebpf_return_type {
+    EBPF_RETURN_TYPE_INTEGER = 0,
+    EBPF_RETURN_TYPE_PTR_TO_MAP_VALUE_OR_NULL,
+    EBPF_RETURN_TYPE_INTEGER_OR_NO_RETURN_IF_SUCCEED,
+} ebpf_return_type_t;
+```
+
+#### `ebpf_helper_function_addresses_t` Struct
+This structure is used to specify the address at which the various helper function implemented by the extension resides. If an eBPF program is JIT compiled, then the generated machine code will have `call` instructions to these addresses. For interpreted mode, the eBPF Execution Engine will invoke the functions at these addresses. The fields of this struct should be set as follows:
+* `helper_function_count`: Number of helper functions implemented by the extension for the given program type.
+* `helper_function_address`: Array of addresses (64-bit unsigned integer) for the helper functions. The addresses must be arranged in the array in the *same order* as the array of helper function prototypes denoted by the `helper_prototype` field in `ebpf_program_info_t` struct.  For the correct execution of eBPF program, the helper function addresses cannot change while a loaded eBPF program is executing.
+
+### 2.2 Program Information NPI Client Attach and Detach Callbacks
+The eBPF Execution Context registers a program information NPI client module with the NMR for every eBPF program that gets loaded. The Execution Context will use the program type GUID of the program as the NPI ID of the client module. And as a result, upon eBPF program load, the associated program information NPI client module will attach with the corresponding program information NPI provider module in the extension. The program information NPI does not have any client or provider dispatch tables. Neither does the `NpiSpecificCharacteristics` from client has any data. So, no special processing is required in the client attach and detach callback handler on the provider module. may decrement the client count. An extension must not unload until there are attached program information NPI clients.
```

Comment:
```suggestion
The eBPF Execution Context registers a program information NPI client module with the NMR for every eBPF program that gets loaded. The Execution Context will use the program type GUID of the program as the NPI ID of the client module. And as a result, upon eBPF program load, the associated program information NPI client module will attach with the corresponding program information NPI provider module in the extension. The program information NPI does not have any client or provider dispatch tables. Neither does the client's `NpiSpecificCharacteristics` have any data. So, no special processing is required in the client attach and detach callback handler on the provider module. WHAT GOES HERE? may decrement the client count. An extension must not unload until there are attached program information NPI clients.
```

grammar.  Also please fill in missing words

Reply:
Fixed.